---
title: "meiosis paper-V3 code"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


```{r cars}
summary(cars)
```


```{r, echo=TRUE, error=TRUE}

.libPaths()
.libPaths( c( "C:/Users/fxy19/Documents/R/win-library/3.6" , .libPaths() ) )

setwd("E:/xueying Fan/meiosis-dec/R running/final-June")
outputDir = getwd()

suppressMessages(library(ggplot2))
suppressMessages(library(Seurat))
suppressMessages(library(ggplot2))
suppressMessages(library(gplots))
suppressMessages(library(plotly))
suppressMessages(library(dplyr))
suppressMessages(library(readxl))
suppressMessages(library(biomaRt))
suppressMessages(library(ggpubr))
suppressMessages(library(reshape2))
suppressMessages(library(pheatmap))
suppressMessages(library(stringr))

sessionInfo()

FGC <- readRDS("E:/xueying Fan/dataset R running/FGC.rds")
lateMeiot <- readRDS("E:/xueying Fan/dataset R running/lateMeiot.rds")
male <- readRDS("E:/xueying Fan/dataset R running/maleMNN.rds")

```

##Fig.S1A, S1B
```{r, echo=TRUE, error=TRUE}
#################Fig.S1A####################
TSNEPlot(FGC, group.by="seurat_clusters",pt.size=2)

#################Fig.S1B####################
FGC@meta.data$age <- c ("5W")
FGC@meta.data[c(grep("7W_",rownames(FGC@meta.data))),7] <- c("7W")
FGC@meta.data[c(grep("8W_",rownames(FGC@meta.data))),7] <- c("8W")
FGC@meta.data[c(grep("10W_",rownames(FGC@meta.data))),7] <- c("10W")
FGC@meta.data[c(grep("11W_",rownames(FGC@meta.data))),7] <- c("11W")
FGC@meta.data[c(grep("12W_",rownames(FGC@meta.data))),7] <- c("12W")
FGC@meta.data[c(grep("18W_",rownames(FGC@meta.data))),7] <- c("18W")
FGC@meta.data[c(grep("20W_",rownames(FGC@meta.data))),7] <- c("20W")
FGC@meta.data[c(grep("23W_",rownames(FGC@meta.data))),7] <- c("23W")
FGC@meta.data[c(grep("24W_",rownames(FGC@meta.data))),7] <- c("24W")
FGC@meta.data[c(grep("26W_",rownames(FGC@meta.data))),7] <- c("26W")

TSNEPlot(FGC, group.by="age",pt.size=2)

```

##Fig.S1C
```{r, echo=TRUE, error=TRUE}

FeaturePlot(FGC, features = c("WT1","COL3A1","COL1A1","VCAN","FOXL2","ALDH1A1","CD68","CD4","VWF","CD34","KIT","DAZL","GSTM1","GABRA3","POU5F1","STRA8","SPO11","ZP3"), reduction = "tsne")

```


##Fig.S1D
```{r, echo=TRUE, error=TRUE}
groupIdentity <- Idents(object = FGC)
nFeature <- FGC@meta.data$nFeature_RNA

featurebox <- data.frame(nFeature,groupIdentity)


fill <- "#9ACD32"
line <- "#2F4F4F"

gg<-ggplot(featurebox)+ geom_boxplot(aes(x = groupIdentity, y = nFeature), fill = fill, colour = line)+ 
  theme_bw()+ 
  geom_jitter(aes(x = groupIdentity, y = nFeature)) 
gg


featurebox$line <- c("germ")
featurebox$line[which(featurebox$groupIdentity %in% c("0","5","8"))] <- "soma"

ggplot(featurebox, aes(x=nFeature, fill=line, color=line)) +
  geom_density(alpha=0.2)

```



##Fig.1A, 1B
```{r, echo=TRUE, error=TRUE}
#################Fig.1A####################
TSNEPlot(lateMeiot, group.by="seurat_clusters",pt.size=2)

#################Fig.1B####################
lateMeiot@meta.data$age <- c ("11W")
lateMeiot@meta.data[c(grep("12W_",rownames(lateMeiot@meta.data))),8] <- c("12W")
lateMeiot@meta.data[c(grep("18W_",rownames(lateMeiot@meta.data))),8] <- c("18W")
lateMeiot@meta.data[c(grep("20W_",rownames(lateMeiot@meta.data))),8] <- c("20W")
lateMeiot@meta.data[c(grep("23W_",rownames(lateMeiot@meta.data))),8] <- c("23W")
lateMeiot@meta.data[c(grep("24W_",rownames(lateMeiot@meta.data))),8] <- c("24W")
lateMeiot@meta.data[c(grep("26W_",rownames(lateMeiot@meta.data))),8] <- c("26W")

TSNEPlot(lateMeiot, group.by="age",pt.size=2)

```


##Fig.1C, 1D
```{r, echo=TRUE, error=TRUE}
########Fig.1C###########
FeaturePlot(lateMeiot, features = c("REC8","SYCE2","PPP1R36","RPL10L","ZP2"), reduction = "tsne")

########Fig.1D###########
VlnPlot(lateMeiot,features = c("DYDC1", "FBXO47", "VCX2","VCX3A","VCX","VCX3B"))

```


##Fig.1E
```{r, echo=TRUE, error=TRUE}
#################Fig.1E####################
marker<- read_excel("E:/xueying Fan/meiosis-dec/new/new figures-2/new file from Susana/lateMeiotAllGenesPadj0.05Pct.1_0.6.xls")

disease<- read_excel("E:/xueying Fan/meiosis-dec/R running/R running knit/Disease.xlsx")



disease.gene <- disease[which(disease$Gene %in% marker$gene),]

disease.gene <- disease.gene[-which(disease.gene$Disease=="Down Syndrome"),]

du.gene <- disease.gene[which(duplicated(disease.gene$'Gene')),]

du.gene <- disease.gene[which(disease.gene$Gene %in% du.gene$Gene),]
uni.gene <- disease.gene[-which(disease.gene$Gene %in% du.gene$Gene),]

genelist <- as.vector(c(uni.gene$'Gene', unique(du.gene$'Gene')))


##################
disease.G <- subset(x = lateMeiot, features = genelist)

## Calculate Mean
groupIdentity <- Idents(object = disease.G)
expr1 <- GetAssayData(object = disease.G, slot = "data")
colnames(expr1) <- groupIdentity


meanDF <- do.call(cbind, lapply(levels(groupIdentity), function(id) {
  groupCounts <- expr1[, colnames(expr1) == id]
  df <- data.frame(c = apply(groupCounts, 1, mean))
  colnames(df) <- id
  return(df)
}))


#############draw heatmap#############################
e <- as.matrix(meanDF[match(genelist,row.names(meanDF)),])

my_palette <- colorRampPalette(c("blue", "white", "red"))(n = 100)

heatmap.2(e, col=my_palette, Rowv = FALSE, 
          trace="none",cexRow=0.2)

```


##Fig.2A
```{r, echo=TRUE, error=TRUE}
########Fig.2A###########

markers.dotplot2 <- c("TUBB2B", "TUBA1A", "TUBB", "TUBA1B", "TUBA3D", "TUBA3C", "TUBA1C","TUBB8","TUBA4A","TUBA4B", "ACTG1","ACTG2","KIF22","KIF9","KIF25","KIF11","KIF20B","KIF9-AS1","KIF21A","KIF1A","KIF18A","KIF5B","KIF13B","DYNLL1","DYNLRB1","DYNLT3","MYO6","MYO1F","MYOM2","MYO5B","MYO1E","MYO7B","MYO10","MYO7A","MYO1H","MYOF","MYO3B","ITGB1","ITGB3BP","ITGAV","ITGA5","ITGA4","ITGA9","ITGA11","CDH2","PCDH11X","PCDH7","CDH1","PCDH15","CDH17","CDH10")

DotPlot(lateMeiot, features = rev(markers.dotplot2), cols = c("blue", "red"), dot.scale = 6)+ RotatedAxis()

```



##Fig.S2A, S2B, S2C
```{r, echo=TRUE, error=TRUE}
#################Fig.S2A####################
TSNEPlot(male, group.by="seurat_clusters",pt.size=2)

#################Fig.S2B####################
TSNEPlot(male, group.by="patientIDs",pt.size=2)

#################Fig.S2C####################
FeaturePlot(male, features = c("VIM","PDGFRA","CD4","AMH","DMRT1","SCML1","TDRG1","OVOL2","NME8"," TXNDC2"), reduction = "tsne")

```


##Fig.3A
```{r, echo=TRUE, error=TRUE}
################Fig.3A###################
female.marker<- read_excel("E:/xueying Fan/meiosis-dec/new/new figure-3/lateMeiotAllGenesPadj0.05Pct.1_0.6.xls")
male.marker<- read_excel("E:/xueying Fan/meiosis-dec/new/new figure-3/maleAllGenes_mnn_Padj0.05Pct.1_0.6.xls")

female.L<- female.marker[female.marker$cluster== "1",]
female.Z<- female.marker[female.marker$cluster== "4",]
female.P<- female.marker[female.marker$cluster== "3",]
female.D<- female.marker[female.marker$cluster== "2",]

male.L <- male.marker[male.marker$cluster== "3",]
male.Z <- male.marker[male.marker$cluster== "1",]
male.eP <- male.marker[male.marker$cluster== "7",]
male.lP <- male.marker[male.marker$cluster== "5",]
male.D <- male.marker[male.marker$cluster== "6",]


L.common <- female.L[which(female.L$'gene' %in% male.L$'gene'),]
female.L.specific <- female.L[-which(female.L$'gene' %in% male.L$'gene'),]
male.L.specific <- male.L[-which(male.L$'gene' %in% female.L$'gene'),]


Z.common <- female.Z[which((female.Z$'gene' %in% male.Z$'gene')&(female.Z$'gene' %in% male.eP$'gene')),]
female.Z.specific <- female.Z[-which((female.Z$'gene' %in% male.Z$'gene')|(female.Z$'gene' %in% male.eP$'gene')),]
male.Z.specific <- male.Z[-which((male.Z$'gene' %in% female.Z$'gene')|(male.Z$'gene' %in% male.eP$'gene')),]
m.ZeP.specific <- male.eP[-which((male.eP$'gene' %in% female.Z$'gene')|(male.eP$'gene' %in% male.Z$'gene')),]

P.common <- female.P[which((female.P$'gene' %in% male.lP$'gene')&(female.P$'gene' %in% male.eP$'gene')),]
m.PeP.specific <- male.eP[-which((male.eP$'gene' %in% female.P$'gene')|(male.eP$'gene' %in% male.lP$'gene')),]
female.P.specific <- female.P[-which((female.P$'gene' %in% male.lP$'gene')|(female.P$'gene' %in% male.eP$'gene')),]
male.lP.specific <- male.lP[-which((male.lP$'gene' %in% female.P$'gene')|(male.lP$'gene' %in% male.eP$'gene')),]

D.common <- female.D[which((female.D$'gene' %in% male.D$'gene')),]
female.D.specific <- female.D[-which((female.D$'gene' %in% male.D$'gene')),]
male.D.specific <- male.D[-which((male.D$'gene' %in% female.D$'gene')),]

nrow(L.common)
nrow(female.L.specific)
nrow(male.L.specific)

nrow(Z.common)
nrow(female.Z.specific)
nrow(male.Z.specific)
nrow(m.ZeP.specific)

nrow(P.common)
nrow(m.PeP.specific)
nrow(female.P.specific)
nrow(male.lP.specific)

nrow(D.common)
nrow(female.D.specific)
nrow(male.D.specific)
```



##Fig.S2D
```{r, echo=TRUE, error=TRUE}
maleMeiosis <- subset(male, ident=c(1,3,5,6,7))
maleMeiosis <- RenameIdents(object = maleMeiosis, 
                            '1' = 'M_Z',
                            '3' = 'M_L', 
                            '5' = 'M_lP',
                            '6' = 'M_D',
                            '7' = 'M_eP')

femaleMeiosis <- subset(lateMeiot, ident=c(1,2,3,4))
femaleMeiosis <- RenameIdents(object = femaleMeiosis,
                          '1' = 'F_L',
                          '4' = 'F_Z',
                          '3' = 'F_P',
                          '2' = 'F_D')

## Combine datasets
maleFemaleMeiosis <- merge(maleMeiosis, y=femaleMeiosis, add.cell.ids = c("M_", "F_"))

## Cells in each dataset
dim(maleMeiosis)
dim(femaleMeiosis)

maleFemaleMeiosis <- FindVariableFeatures(object = maleFemaleMeiosis, 
                         mean.function = ExpMean, 
                         dispersion.function = LogVMR,
                         nfeatures = 2500)
length(x = VariableFeatures(maleFemaleMeiosis))

#####scaling the data
maleFemaleMeiosis <- ScaleData(object = maleFemaleMeiosis)

#####PCA
maleFemaleMeiosis <- RunPCA(object = maleFemaleMeiosis, 
              features = VariableFeatures(maleFemaleMeiosis), 
              do.print = TRUE, 
              pcs.print = 1:5, 
              genes.print = 5)

VizDimLoadings(object = maleFemaleMeiosis, 
               dims = 1:2, 
               reduction = "pca")

DimPlot(object = maleFemaleMeiosis, reduction = "pca")

ElbowPlot(object = maleFemaleMeiosis)

## Print the percenatge of the standard deviation of each PC as a 
## fraction of the total standard deviation od the first 20 PCs
totalSdev <- sum(Stdev(object = maleFemaleMeiosis, reduction = "pca"))
print(Stdev(object = maleFemaleMeiosis, reduction = "pca")/totalSdev)

maleFemaleMeiosis <- FindNeighbors(maleFemaleMeiosis,
                                   reduction = "pca",
                                   dims = 1:11)
maleFemaleMeiosis <- FindClusters(object = maleFemaleMeiosis, 
                                  resolution = 0.6)

maleFemaleMeiosis <- RunTSNE(object = maleFemaleMeiosis, 
                             dims = 1:11)
maleFemaleMeiosis <- RunUMAP(object = maleFemaleMeiosis, 
                             dims = 1:11)

DimPlot(object = maleFemaleMeiosis, 
        reduction = "tsne")

###Get gender of cells and plot
gender <- sapply(colnames(x = maleFemaleMeiosis), function(x){
  strsplit(x, split = "__", fixed = TRUE)[[1]][1]
})
maleFemaleMeiosis$gender <- gender
DimPlot(object = maleFemaleMeiosis, 
             reduction = "tsne",
             group.by = "gender")
```



##Fig.S2E
```{r, echo=TRUE, error=TRUE}
set.seed(42)
sampledCellIDs <- sample(x = colnames(maleMeiosis), size = 160, replace = FALSE)
maleMeiosisSubsmpl <- subset(x = maleMeiosis, cells=sampledCellIDs)
maleFemaleMeiosisEqualized <- merge(maleMeiosisSubsmpl, y=femaleMeiosis, add.cell.ids = c("M_", "F_"))
## Add the M_ in the cell ID
sampledIDs <- paste0("M__",sampledCellIDs)
DimPlot(object = maleFemaleMeiosis, 
             reduction = "tsne",
             cells.highlight = sampledIDs)

###Find Variable Genes
maleFemaleMeiosisEqualized <- FindVariableFeatures(object = maleFemaleMeiosisEqualized, 
                         mean.function = ExpMean, 
                         dispersion.function = LogVMR,
                         nfeatures = 2500)
length(x = VariableFeatures(maleFemaleMeiosisEqualized))

topAveExpr = HVFInfo(maleFemaleMeiosisEqualized)[HVFInfo(maleFemaleMeiosisEqualized)[,1]>3, ]

##Scaling the data
maleFemaleMeiosisEqualized <- ScaleData(object = maleFemaleMeiosisEqualized)
##PCA
maleFemaleMeiosisEqualized <- RunPCA(object = maleFemaleMeiosisEqualized, 
                                     features = VariableFeatures(maleFemaleMeiosisEqualized), 
                                     do.print = TRUE, 
                                     pcs.print = 1:5, 
                                     genes.print = 5)

##Vizualize PCA loadings
VizDimLoadings(object = maleFemaleMeiosisEqualized, 
               dims = 1:2, 
               reduction = "pca")

DimPlot(object = maleFemaleMeiosisEqualized, reduction = "pca")

ElbowPlot(object = maleFemaleMeiosisEqualized)

## Print the percenatge of the standard deviation of each PC as a 
## fraction of the total standard deviation od the first 20 PCs
totalSdev <- sum(Stdev(object = maleFemaleMeiosisEqualized, reduction = "pca"))
print(Stdev(object = maleFemaleMeiosisEqualized, reduction = "pca")/totalSdev)

###tSNEs and UMAP
maleFemaleMeiosisEqualized <- RunTSNE(object = maleFemaleMeiosisEqualized, 
                             dims = 1:11)
maleFemaleMeiosisEqualized <- RunUMAP(object = maleFemaleMeiosisEqualized, 
                             dims = 1:11)
DimPlot(object = maleFemaleMeiosisEqualized, 
        reduction = "tsne")

saveRDS(maleFemaleMeiosisEqualized, file = "femalemale.rds")
```


##Fig.3B, 3C,S3A
```{r, echo=TRUE, error=TRUE}

maleFemaleMeiosisEqualized <- readRDS("E:/xueying Fan/meiosis-dec/R running/R running-april/femalemale.rds")

################Fig.3B###################
DimPlot(object = maleFemaleMeiosisEqualized, 
        reduction = "tsne")

################Fig.3C###################
FeaturePlot(maleFemaleMeiosisEqualized, features = c("TEX19","PRDM9","SPO11","MEIOB","BRDT","BAZ2B","AURKA","BTG4","H1FOO","FRMD3"), reduction = "tsne")

################Fig.S3A###################
FeaturePlot(maleFemaleMeiosisEqualized, features = c("PRKACB","HIST1H1E","DMRTA2","PRDM1","XIST","KMT2A","PADI6","WEE2","DNMT3A","TET2"), reduction = "tsne")

FeaturePlot(maleFemaleMeiosisEqualized, features = c("GAGE2A","PAGE1","STAT4","TDRG1","POU5F2","PROK2","PLK1","TEKT4","GK2","IZUMO4"), reduction = "tsne")

```


##Fig.S3C
```{r, echo=TRUE, error=TRUE}
################Fig.S3B###################

female.L.LOC <- female.L[grep("LOC", female.L$gene),] 
female.L.LINC <- female.L[grep("LINC", female.L$gene),] 

female.Z.LOC <- female.Z[grep("LOC", female.Z$gene),] 
female.Z.LINC <- female.Z[grep("LINC", female.Z$gene),] 

female.P.LOC <- female.P[grep("LOC", female.P$gene),] 
female.P.LINC <- female.P[grep("LINC", female.P$gene),]

female.D.LOC <- female.D[grep("LOC", female.D$gene),] 
female.D.LINC <- female.D[grep("LINC", female.D$gene),]

male.L.LOC <- male.L[grep("LOC", male.L$gene),] 
male.L.LINC <- male.L[grep("LINC", male.L$gene),]

male.Z.LOC <- male.Z[grep("LOC", male.Z$gene),] 
male.Z.LINC <- male.Z[grep("LINC", male.Z$gene),]

male.eP.LOC <- male.eP[grep("LOC", male.eP$gene),] 
male.eP.LINC <- male.eP[grep("LINC", male.eP$gene),]

male.lP.LOC <- male.lP[grep("LOC", male.lP$gene),] 
male.lP.LINC <- male.lP[grep("LINC", male.lP$gene),]

male.D.LOC <- male.D[grep("LOC", male.D$gene),] 
male.D.LINC <- male.D[grep("LINC", male.D$gene),]

#######################################################################
F.L <- c(6,6,615)
F.Z <- c(12,2,248)
F.P <- c(6,2,77)
F.D <- c(11,8,680)
M.L <- c(8,1,1410)
M.Z <- c(20,19,2343)
M.EP <- c(79,56,2787)
M.LP <- c(108,51,2822)
M.D <- c(79,43,2428)

summary.loc <- data.frame(F.L, F.Z, F.P, F.D, M.L, M.Z, M.EP, M.LP, M.D)
rownames(summary.loc) <- c("LOC", "LINC","other")

################pie chart################

par(mfrow=(c(2,5)))
pie(summary.loc$F.L, labels = rownames(summary.loc), main="F.L")
pie(summary.loc$F.Z, labels = rownames(summary.loc), main="F.Z")
pie(summary.loc$F.P, labels = rownames(summary.loc), main="F.P")
pie(summary.loc$F.D, labels = rownames(summary.loc), main="F.D")
pie(summary.loc$M.L, labels = rownames(summary.loc), main="M.L")
pie(summary.loc$M.Z, labels = rownames(summary.loc), main="M.Z")
pie(summary.loc$M.EP, labels = rownames(summary.loc), main="M.EP")
pie(summary.loc$M.LP, labels = rownames(summary.loc), main="M.LP")
pie(summary.loc$M.D, labels = rownames(summary.loc), main="M.D")

```



##Fig.3D
```{r, echo=TRUE, error=TRUE}
################Fig.3D###################
###find DEGs between sex
###Get gender of cells and plot
gender <- sapply(colnames(x = maleFemaleMeiosisEqualized), function(x){
  strsplit(x, split = "__", fixed = TRUE)[[1]][1]
})
maleFemaleMeiosisEqualized$gender <- gender

DEGs<- FindMarkers(object=maleFemaleMeiosisEqualized,
                   ident.1 = colnames(maleFemaleMeiosisEqualized[,which(maleFemaleMeiosisEqualized@meta.data$gender=="M")]), ident.2 =  colnames(maleFemaleMeiosisEqualized[,which(maleFemaleMeiosisEqualized@meta.data$gender=="F")]))

nrow(DEGs)

library(EnhancedVolcano)
EnhancedVolcano(DEGs,
    lab = rownames(DEGs),
    x = 'avg_logFC',
    y = 'p_val_adj',
    pCutoff = 0.001,
    FCcutoff = 1.5,
    xlim = c(-5, 8),
    pLabellingCutoff = NULL)

DEGs.filtered1 <- DEGs[(DEGs$p_val_adj<0.001 & DEGs$avg_logFC>1.5),]
nrow(DEGs.filtered1) 

DEGs.filtered2 <- DEGs[(DEGs$p_val_adj<0.001 & DEGs$avg_logFC< -1.5),]
nrow(DEGs.filtered2) 


```



##Fig.3E 3F
```{r, echo=TRUE, error=TRUE}

DEGs.filtered1$GENE <- rownames(DEGs.filtered1)
top100.male <- DEGs.filtered1 %>% top_n(n = 100, wt = avg_logFC)

DEGs.filtered2$GENE <- rownames(DEGs.filtered2)
top100.female <- DEGs.filtered2 %>% top_n(n = -100, wt = avg_logFC)

top.female.genes <- as.vector(top100.female$GENE)
top.male.genes <- as.vector(top100.male$GENE)

#################Fig.3E####################
#######################plot heatmap of female top100 ######################
Top.female<- subset(x = maleFemaleMeiosisEqualized, features = top.female.genes)

## Calculate Mean
groupIdentity <- Idents(object = Top.female)
expr1 <- GetAssayData(object = Top.female, slot = "data")
colnames(expr1) <- groupIdentity


meanDF <- do.call(cbind, lapply(levels(groupIdentity), function(id) {
    groupCounts <- expr1[, colnames(expr1) == id]
    df <- data.frame(c = apply(groupCounts, 1, mean))
    colnames(df) <- id
    return(df)
}))


#############draw heatmap#############################
rv <- genefilter::rowVars(meanDF)
idx <- order(-rv)
e <- as.matrix(meanDF)

heatmap.2(e[idx, ], col=my_palette, 
    trace="none",cexRow=0.2)

#################Fig.3F####################
#######################plot heatmap of male top100 ######################
Top.male<- subset(x = maleFemaleMeiosisEqualized, features = top.male.genes)

## Calculate Mean
groupIdentity <- Idents(object = Top.male)
expr1 <- GetAssayData(object = Top.male, slot = "data")
colnames(expr1) <- groupIdentity


meanDF <- do.call(cbind, lapply(levels(groupIdentity), function(id) {
    groupCounts <- expr1[, colnames(expr1) == id]
    df <- data.frame(c = apply(groupCounts, 1, mean))
    colnames(df) <- id
    return(df)
}))


#############draw heatmap#############################
rv <- genefilter::rowVars(meanDF)
idx <- order(-rv)
e <- as.matrix(meanDF)

heatmap.2(e[idx, ], col=my_palette, 
    trace="none",cexRow=0.2)

```


##Fig.S3B
```{r, echo=TRUE, error=TRUE}
################Fig.S3B###################
DotPlot(male, features = rev(markers.dotplot2), cols = c("blue", "red"), dot.scale = 6)+ RotatedAxis()

```


##Fig.4A
```{r, echo=TRUE, error=TRUE}
library(RColorBrewer)
library(biomaRt)
mart <- useMart(biomart="ensembl", dataset="hsapiens_gene_ensembl")
chr.DEGs <- getBM(attributes = c('hgnc_symbol', 'chromosome_name'),
                                filters = 'hgnc_symbol', 
                                values = rownames(DEGs), 
                                mart = mart)

DEGs2 <- DEGs[-which(rownames(DEGs) %in% chr.DEGs$hgnc_symbol),]
chr.DEGs2 <- getBM(attributes = c('hgnc_symbol', 'chromosome_name','external_synonym'),
                                filters = 'external_synonym', 
                                values = rownames(DEGs2), 
                                mart = mart)
chr.DEGs2 $hgnc_symbol <- chr.DEGs2$external_synonym
chr.DEGs2 <- chr.DEGs2[,-3]

chr.DEGs <- rbind(chr.DEGs,chr.DEGs2)
chr.DEGs <- chr.DEGs[-c(grep("CHR_",chr.DEGs$chromosome_name)),]

################################female DEGs###################

DEGs.female <- chr.DEGs[which(chr.DEGs$hgnc_symbol %in% rownames(DEGs.filtered2)),]
mytable <- table(DEGs.female$chromosome_name)
mytable <- as.data.frame(mytable)
mytable$percentage <- (mytable$Freq/sum(mytable$Freq))*100
colourCount = length(mytable$Var1)
getPalette = colorRampPalette(brewer.pal(9, "Set1"))
ggplot(mytable, aes(x=Var1, y=percentage, fill=Var1))+
geom_bar(stat="identity", color="black")+ 
  scale_fill_manual(values = getPalette(colourCount)) +
  theme(legend.position="bottom") +
  guides(fill=guide_legend(nrow=2))+ylim(0, 20)

################################male DEGs###################

DEGs.male <- chr.DEGs[which(chr.DEGs$hgnc_symbol %in% rownames(DEGs.filtered1)),]
mytable <- table(DEGs.male$chromosome_name)
mytable <- as.data.frame(mytable)
mytable$percentage <- (mytable$Freq/sum(mytable$Freq))*100
colourCount = length(mytable$Var1)
getPalette = colorRampPalette(brewer.pal(9, "Set1"))
ggplot(mytable, aes(x=Var1, y=percentage, fill=Var1))+
geom_bar(stat="identity", color="black")+ 
  scale_fill_manual(values = getPalette(colourCount)) +
  theme(legend.position="bottom") +
  guides(fill=guide_legend(nrow=2))+ylim(0, 20)

```


##Fig.4B
```{r, echo=TRUE, error=TRUE}
#########################map selected genes on X chromosome#####################
x.degs <- chr.DEGs[which(chr.DEGs$chromosome_name=="X"),]
xgenes.femalemale <- DEGs[which(rownames(DEGs) %in% x.degs$hgnc_symbol),]

xgenes.femalemale <- xgenes.femalemale[which(xgenes.femalemale$p_val_adj<0.001),]
xgenes.femalemale <- xgenes.femalemale[which(xgenes.femalemale$avg_logFC>1.5|xgenes.femalemale$avg_logFC< -1.5),]

library(regioneR)
library(IRanges)
ensembl <- useEnsembl(biomart="ensembl", dataset="hsapiens_gene_ensembl", host = "useast.ensembl.org")
x.genes <- getBM(attributes = c('hgnc_symbol', 'chromosome_name', 'start_position', 'end_position'),
                                filters = 'hgnc_symbol', 
                                values = rownames(xgenes.femalemale), 
                                mart = ensembl)

x.genes2 <- xgenes.femalemale[-which(rownames(xgenes.femalemale) %in% x.genes$hgnc_symbol),]
x.genes2 <- getBM(attributes = c('hgnc_symbol', 'chromosome_name', 'start_position', 'end_position','external_synonym'),
                                filters = 'external_synonym', 
                                values = rownames(x.genes2), 
                                mart = ensembl)
x.genes2 $hgnc_symbol <- x.genes2$external_synonym
x.genes2 <- x.genes2[,-5]

x.genes <- rbind(x.genes,x.genes2)
xgenes.femalemale <- xgenes.femalemale[match(x.genes$hgnc_symbol,row.names(xgenes.femalemale)),]
x.genes$avg_logFC <- xgenes.femalemale$avg_logFC
x.genes$p_val_adj <-xgenes.femalemale$p_val_adj

genes <- GenomicRanges::GRanges(seqnames=x.genes[,2], ranges=IRanges::IRanges(start=x.genes[,3], end=x.genes[,4]))
values(genes) <- DataFrame(hgnc_symbol =x.genes$hgnc_symbol,avg_logFC = x.genes$avg_logFC)
seqlevelsStyle(genes) <- "UCSC"
head(genes)

library(karyoploteR)

fc.ymax <- ceiling(max(abs(range(genes$'avg_logFC'))))
fc.ymin <- -fc.ymax


sel.data <- genes[which(genes$'hgnc_symbol'%in% c('AKAP4','FMR1NB','SPANXN5','LINC00633','SPANXN3','CYLC1','CXorf65','HYPM',  'PAGE1','SPANXN4','FTH1P18','PPP1R2P9','CXorf66','CT83','AKAP14','ZCCHC13','GAGE2A','PAGE4','UBE2DNL','GAGE1','SSX3',      'TMEM31','FAM47B','RPS26P11','CPXCR1','ZNF645','ARL13A','H2BFM','SSX1','TAF9B','SLC25A5','PGK1', 'ATRX','RHOXF1','FGF13','KLHL13','GPM6B','COL4A5','SMS','XIST')),]
head(sel.data)

plot.new()
points.top <- 0.8
kp <- plotKaryotype(genome="hg38", chromosomes="chrX",plot.type=1)
kpplot <- kpAddBaseNumbers(kp)
kpAxis(kp, ymax=5, ymin=-5,numticks = 5,tick.pos = c(-5, -1.5, 0, 1.5,5), cex=1.8,r1=points.top)
point.colors <- colByValue(genes$'avg_logFC', colors = c("green", "black","red"), min = -1.5, max=1.5)
kpPoints(kp, genes, y=genes$'avg_logFC',ymax=fc.ymax, ymin=fc.ymin, col=point.colors, cex=0.8,r1=points.top)
gene.mean <- (end(sel.data) - start(sel.data))/2+start(sel.data)
kpSegments(kp, chr="chrX", x0=gene.mean, x1=gene.mean,y0=sel.data$'avg_logFC', y1=fc.ymax, ymax=fc.ymax, ymin=fc.ymin, r1=points.top)
kpPlotMarkers(kp, sel.data, labels = sel.data$'hgnc_symbol', text.orientation = "vertical",r0=points.top)



```


##Fig.S4A
```{r, echo=TRUE, error=TRUE}
##import x-linked escape gene list
Xesc<- read_excel("E:/xueying Fan/meiosis-dec/R running/R running knit/Eseacpees with reference.new.xlsx")
Xesclist <- Xesc$`Gene Name`

xgenes.female <- xgenes.femalemale[(xgenes.femalemale$p_val_adj<0.001 & xgenes.femalemale$avg_logFC< -1.5),]
nrow(xgenes.female) 

xgenes.male <- xgenes.femalemale[(xgenes.femalemale$p_val_adj<0.001 & xgenes.femalemale$avg_logFC>1.5),]
nrow(xgenes.male) 

###############################heatmap of female X-linked genes##############################################
## Calculate female Mean
x.female<- subset(maleFemaleMeiosisEqualized, features = rownames(xgenes.female))
expr1 <- GetAssayData(object = x.female, slot = "data")

groupIdentity <- Idents(object = x.female)
colnames(expr1) <- groupIdentity


meanDF <- do.call(cbind, lapply(unique(groupIdentity), function(id) {
  groupCounts <- expr1[, colnames(expr1) == id]
  df <- data.frame(c = apply(groupCounts, 1, mean))
  colnames(df) <- id
  return(df)
}))

##female annotation
xgenes.female$type <- c("pro")

xgenes.female.esc <- which(rownames(xgenes.female) %in% Xesclist)
xgenes.female.esc <- xgenes.female[xgenes.female.esc,]
nrow(xgenes.female.esc)

esc <- rownames(xgenes.female.esc)
xgenes.female[esc,6] <- c("esc")

anno <- as.data.frame (xgenes.female[,6])
rownames(anno) <-rownames(xgenes.female)
names(anno) <- "X.types"

meanDF <- as.matrix(meanDF)

my_palette <- colorRampPalette(c("blue", "white", "red"))(n = 100)
pheatmap(mat = meanDF, annotation_row = anno, show_rownames = F, scale = "none", color = my_palette)




###############################heatmap of male X-linked genes##############################################

## Calculate male Mean
x.male<- subset(maleFemaleMeiosisEqualized, features = rownames(xgenes.male))
expr1 <- GetAssayData(object = x.male, slot = "data")

groupIdentity <- Idents(object = x.male)
colnames(expr1) <- groupIdentity


meanDF <- do.call(cbind, lapply(unique(groupIdentity), function(id) {
  groupCounts <- expr1[, colnames(expr1) == id]
  df <- data.frame(c = apply(groupCounts, 1, mean))
  colnames(df) <- id
  return(df)
}))

##female annotation
xgenes.male$type <- c("pro")

xgenes.male.esc <- which(rownames(xgenes.male) %in% Xesclist)
xgenes.male.esc <- xgenes.male[xgenes.male.esc,]
nrow(xgenes.male.esc)

esc <- rownames(xgenes.male.esc)
xgenes.male[esc,6] <- c("esc")

anno <- as.data.frame (xgenes.male[,6])
rownames(anno) <-rownames(xgenes.male)
names(anno) <- "X.types"

meanDF <- as.matrix(meanDF)

pheatmap(mat = meanDF, annotation_row = anno, show_rownames = F, scale = "none", color = my_palette)

```



##Fig.4C
```{r, echo=TRUE, error=TRUE}

VlnPlot(lateMeiot,features = c("XIST","YY1","SPEN","EZH2","H2AFY","TSIX"))

```



## SNPs filtering and table generation
```{r, echo=TRUE, error=TRUE}

chr7.cov5<- read_excel("E:/xueying Fan/meiosis-dec/new/SNPs/05-05-2020/05-05-2020/allele_depth_table_filter_cells_depth_reordered_gene_names_chr7_masked.xls")
chrx.cov5<- read_excel("E:/xueying Fan/meiosis-dec/new/SNPs/05-05-2020/05-05-2020/allele_depth_table_filter_cells_depth_reordered_gene_names_chrX_masked.xls")

########################################split chx.cov5###########################################
chrx.cov5.2<- chrx.cov5[,10:236]

x <- apply(chrx.cov5.2,2,as.list)

for(i in 1:227){
   df<- t(as.data.frame(strsplit(as.character(x[[i]]), '|', fixed=TRUE)))
    x[[i]] <- df
}
chrx.cov5.2 <- as.data.frame(x, stringsAsFactors=FALSE)
rownames(chrx.cov5.2) <-rownames(chrx.cov5)
 chrx.cov5.2 <- data.frame(chrx.cov5[,c(1:9)], chrx.cov5.2)
 chrx.cov5.2[,c(10:463)] <- lapply(chrx.cov5.2[,c(10:463)], as.integer)

#####filter
y1 <- chrx.cov5.2[,grep(".1$",colnames(chrx.cov5.2))]
y2 <- chrx.cov5.2[,grep(".2$",colnames(chrx.cov5.2))]


y3 <- as.data.frame(str_match(colnames(y1), "(.*)_(.*)_CL_(.*)$"))
y3$cluster <- stringr::str_replace(y3$V4, '.\\d+', '')
y3$V4 <- c("ref")
colnames(y3)<- c("id","embryo","cell","allele","cluster")

y4 <- as.data.frame(str_match(colnames(y2), "(.*)_(.*)_CL_(.*)$"))
y4$cluster <- stringr::str_replace(y4$V4, '.\\d+', '')
y4$V4 <- c("alt")
colnames(y4)<- c("id","embryo","cell","allele","cluster")

meta <- rbind(y3,y4)

  

meta.embryo <- do.call(cbind,lapply(unique(meta$embryo), function(x){
    
     y3<-y1[,grep(as.character(x),colnames(y1))]
    y4<-y2[,grep(as.character(x),colnames(y2))]
    
    for(i in 1:nrow(y3)){
    
      y5 <- y3[i,]
      y6 <- y4[i,]
      y7 <-y5[,which(y5>5)]
      y8 <- y6[,which(colnames(y6) %in% sub(".1$", ".2", colnames(y7)))]
      
      if (all(y5<5)| all(y6<5)| all(y8<5)) 
    { y3[i,] <- NA
      y4[i,] <- NA}
      
    }
    
    y3 <- as.data.frame(y3)
    colnames(y3)<- sub(".1$", "", colnames(y3))
    
  return(y3)
    
  }))
 
meta.embryo2<-  do.call(cbind,lapply(unique(meta$embryo), function(x){
  
  y3 <- meta.embryo[,grep(as.character(x), colnames(meta.embryo))]
  y4 <- as.data.frame(y3[,1])
  y4[which(!is.na(y4)),1] <- 1
  names(y4)<-x
  return(y4)
      
    } )) 

meta.embryo2 <- data.frame(chrx.cov5[,c(1:9)], meta.embryo2 )

for(i in 1:nrow(y1)){
  y3 <- colnames(meta.embryo2[,10:20])[which(meta.embryo2[i,10:20]=='1')]
   y4 <- meta[which(meta$embryo %in% y3),1]
   y5 <- y1[i,which(colnames(y1)%in% y4)]
   y6 <- colnames(y5)[which(y5>0)]
   y7 <- sub(".1$", ".2", y6)
   y8 <- y2[i,y7]
   
   if(length(which(y8==0))>30){
     chrx.cov5[i,] <- NA
     }
   }

chrx.cov5 <- chrx.cov5[complete.cases(chrx.cov5), ]
dim(chrx.cov5)

meta.embryo2 <- meta.embryo2[which(meta.embryo2$coord %in% chrx.cov5$coord),]

write.table(chrx.cov5, file = paste0(outputDir, "/chrx.cov5-30cells.tsv"), sep = "\t", quote = FALSE)
write.table(meta.embryo2, file = paste0(outputDir, "/chrx.cov5-30cells.meta.tsv"), sep = "\t", quote = FALSE)


########################################split ch7.cov5###########################################
chr7.cov5.2<- chr7.cov5[,9:235]

x <- apply(chr7.cov5.2,2,as.list)

for(i in 1:227){
   df<- t(as.data.frame(strsplit(as.character(x[[i]]), '|', fixed=TRUE)))
    x[[i]] <- df
}
chr7.cov5.2 <- as.data.frame(x, stringsAsFactors=FALSE)
rownames(chr7.cov5.2) <-rownames(chr7.cov5)
 chr7.cov5.2 <- data.frame(chr7.cov5[,c(1:8)], chr7.cov5.2)
 chr7.cov5.2[,c(9:462)] <- lapply(chr7.cov5.2[,c(9:462)], as.integer)

#####filter
y1 <- chr7.cov5.2[,grep(".1$",colnames(chr7.cov5.2))]
y2 <- chr7.cov5.2[,grep(".2$",colnames(chr7.cov5.2))]


y3 <- as.data.frame(str_match(colnames(y1), "(.*)_(.*)_CL_(.*)$"))
y3$cluster <- stringr::str_replace(y3$V4, '.\\d+', '')
y3$V4 <- c("ref")
colnames(y3)<- c("id","embryo","cell","allele","cluster")

y4 <- as.data.frame(str_match(colnames(y2), "(.*)_(.*)_CL_(.*)$"))
y4$cluster <- stringr::str_replace(y4$V4, '.\\d+', '')
y4$V4 <- c("alt")
colnames(y4)<- c("id","embryo","cell","allele","cluster")

meta <- rbind(y3,y4)

  

meta.embryo <- do.call(cbind,lapply(unique(meta$embryo), function(x){
    
     y3<-y1[,grep(as.character(x),colnames(y1))]
    y4<-y2[,grep(as.character(x),colnames(y2))]
    
    for(i in 1:nrow(y3)){
    
      y5 <- y3[i,]
      y6 <- y4[i,]
      y7 <-y5[,which(y5>5)]
      y8 <- y6[,which(colnames(y6) %in% sub(".1$", ".2", colnames(y7)))]
      
      if (all(y5<5)| all(y6<5)| all(y8<5)) 
    { y3[i,] <- NA
      y4[i,] <- NA}
      
    }
    
    y3 <- as.data.frame(y3)
    colnames(y3)<- sub(".1$", "", colnames(y3))
    
  return(y3)
    
  }))
 
meta.embryo2<-  do.call(cbind,lapply(unique(meta$embryo), function(x){
  
  y3 <- meta.embryo[,grep(as.character(x), colnames(meta.embryo))]
  y4 <- as.data.frame(y3[,1])
  y4[which(!is.na(y4)),1] <- 1
  names(y4)<-x
  return(y4)
      
    } )) 

meta.embryo2 <- data.frame(chr7.cov5[,c(1:8)], meta.embryo2 )

for(i in 1:nrow(y1)){
  y3 <- colnames(meta.embryo2[,9:19])[which(meta.embryo2[i,9:19]=='1')]
   y4 <- meta[which(meta$embryo %in% y3),1]
   y5 <- y1[i,which(colnames(y1)%in% y4)]
   y6 <- colnames(y5)[which(y5>0)]
   y7 <- sub(".1$", ".2", y6)
   y8 <- y2[i,y7]
   
   if(length(which(y8==0))>30){
     chr7.cov5[i,] <- NA
     }
   }

chr7.cov5 <- chr7.cov5[complete.cases(chr7.cov5), ]
dim(chr7.cov5)

meta.embryo2 <- meta.embryo2[which(meta.embryo2$coord %in% chr7.cov5$coord),]

write.table(chr7.cov5, file = paste0(outputDir, "/chr7.cov5-30cells.tsv"), sep = "\t", quote = FALSE)
write.table(meta.embryo2, file = paste0(outputDir, "/chr7.cov5-30cells.meta.tsv"), sep = "\t", quote = FALSE)


##################################organize the final table################################################
chrx.cov5 <- read.delim("E:/xueying Fan/meiosis-dec/R running/final-June/chrx.cov5-30cells.tsv")
chr7.cov5 <- read.delim("E:/xueying Fan/meiosis-dec/R running/final-June/chr7.cov5-30cells.tsv")

chr7.cov5 <- data.frame(chr7.cov5[,1:2],chr7.cov5$hgnc, chr7.cov5[,3:235])
colnames(chr7.cov5)[3] <- "hgnc"

## merge chr7 and chrx read counts table
chr7.cov5<- as.matrix(chr7.cov5)
chrx.cov5<- as.matrix(chrx.cov5)
tableSNP <-as.data.frame(rbind(chrx.cov5,chr7.cov5)) 

##
embryox<- read_excel("E:/xueying Fan/meiosis-dec/new/SNPs/05-05-2020/per_embryo_het_calls_cov5_chrX_masked.xls")
embryo7<- read_excel("E:/xueying Fan/meiosis-dec/new/SNPs/05-05-2020/per_embryo_het_calls_cov5_chr7_masked.xls")

embryox <- data.frame(embryox[,1:2],embryox[,40],embryox[,3:39])
embryo7 <- data.frame(embryo7[,1:2],embryo7[,40],embryo7[,3:39])

embryox <- as.matrix(embryox)
embryo7 <- as.matrix(embryo7)

embryox <- embryox[which(embryox[,2] %in% tableSNP$coord),]
embryo7 <- embryo7[which(embryo7[,2] %in% tableSNP$coord),]

tmbryo <- as.data.frame(rbind(embryox,embryo7)) 

tmbryo <- tmbryo[match(tableSNP[,2],tmbryo[,2]),]

tmbryo <- tmbryo[,-c(1:7)]

all <- cbind(tableSNP,tmbryo)

write.table(all, file = paste0(outputDir, "/finalmerged_SNPs.cov5-30cells.tsv"), sep = "\t", quote = FALSE)


```


##Fig.S4B
```{r, echo=TRUE, error=TRUE}
##################################################
chrx.cov5.meta <- read.table("chrx.cov5-30cells.meta.tsv", sep = '\t', header = TRUE)
chr7.cov5.meta <- read.table("chr7.cov5-30cells.meta.tsv", sep = '\t', header = TRUE)

meta.xpro <- chrx.cov5.meta[-which(chrx.cov5.meta$hgnc %in% Xesclist),]
meta.xesc <- chrx.cov5.meta[which(chrx.cov5.meta$hgnc %in% Xesclist),]

chr7 <- as.data.frame(colSums(chr7.cov5.meta[,9:19],na.rm=TRUE))
chrX.esc <- as.data.frame(colSums(meta.xesc[,10:20],na.rm=TRUE))
chrX.pro <- as.data.frame(colSums(meta.xpro[,10:20],na.rm=TRUE))

SNPtable <- cbind(chr7, chrX.esc, chrX.pro)
colnames(SNPtable) <- c("chr7", "chrX.esc", "chrX.pro")
SNPtable$cell <- rownames(SNPtable)
SNPtable <- melt(SNPtable,ID="cell")


ggplot(SNPtable, aes(x=cell, y=value, fill=variable))+
 geom_bar(stat="identity", color="black",position=position_dodge())+
  scale_fill_brewer(palette="Set2")+theme_bw()

```


##Fig.4D
```{r, echo=TRUE, error=TRUE}

chrx.cov5<- read.table("chrx.cov5-30cells.tsv", sep = '\t', header = TRUE)
chr7.cov5<- read.table("chr7.cov5-30cells.tsv", sep = '\t', header = TRUE)

chrx.cov5.meta <- read.table("chrx.cov5-30cells.meta.tsv", sep = '\t', header = TRUE)
chr7.cov5.meta <- read.table("chr7.cov5-30cells.meta.tsv", sep = '\t', header = TRUE)

meta.xpro <- chrx.cov5.meta[-which(chrx.cov5.meta$hgnc %in% Xesclist),]
meta.xesc <- chrx.cov5.meta[which(chrx.cov5.meta$hgnc %in% Xesclist),]

####split ch7.cov5
chr7.cov5.2<- chr7.cov5[,9:235]

x <- apply(chr7.cov5.2,2,as.list)

for(i in 1:227){
   df<- t(as.data.frame(strsplit(as.character(x[[i]]), '|', fixed=TRUE)))
    x[[i]] <- df
}
chr7.cov5.2 <- as.data.frame(x, stringsAsFactors=FALSE)
rownames(chr7.cov5.2) <-rownames(chr7.cov5)
 chr7.cov5.2 <- data.frame(chr7.cov5[,c(1:8,236)], chr7.cov5.2)
 chr7.cov5.2[,c(10:463)] <- lapply(chr7.cov5.2[,c(10:463)], as.integer)
 
####split chx.cov5
 xnorm<- chrx.cov5[,10:236]

x <- apply(xnorm,2,as.list)

for(i in 1:227){
  df<- t(as.data.frame(strsplit(as.character(x[[i]]), '|', fixed=TRUE)))
  x[[i]] <- df
}
xnorm <- as.data.frame(x, stringsAsFactors=FALSE)
rownames(xnorm) <-rownames(chrx.cov5)
xnorm<- data.frame(chrx.cov5[,c(1:9)], xnorm)
xnorm[,c(10:463)] <- lapply(xnorm[,c(10:463)], as.integer)

###################################chr7#####################################
chr7.ref <- chr7.cov5.2[,grep(".1$", colnames(chr7.cov5.2))]
chr7.alt <- chr7.cov5.2[,grep(".2$", colnames(chr7.cov5.2))]

chr7.pct <- chr7.ref+chr7.alt

for (i in 1:nrow(chr7.pct)) {
  y1 <- chr7.pct[i,]
  y2 <- which(y1 <15)
  chr7.ref[i,y2] <- NA
  chr7.alt[i,y2] <- NA
}

chr7.ref <- log(chr7.ref+1)
chr7.alt <- log(chr7.alt+1)


##### 
y4 <- as.data.frame(str_match(colnames(chr7.ref), "(.*)_(.*)_CL_(.*)$"))
y4$cluster <- stringr::str_replace(y4$V4, '.\\d+', '')
y4$V4 <- c("ref")
colnames(y4)<- c("id","embryo","cell","allele","cluster")

y5 <- as.data.frame(str_match(colnames(chr7.alt), "(.*)_(.*)_CL_(.*)$"))
y5$cluster <- stringr::str_replace(y5$V4, '.\\d+', '')
y5$V4 <- c("alt")
colnames(y5)<- c("id","embryo","cell","allele","cluster")

meta <- rbind(y4,y5)

  snp.chr7.cov5 <- do.call(rbind,lapply(1:nrow(chr7.ref), function(x){
   
   y3 <- chr7.ref[x,]
   y4<- chr7.alt[x,]
   y5 <- as.data.frame(cbind(colSums(y3),colSums(y4)))
   colnames(y5) <- c("ref","alt")
   y5 <- data.frame(y5,meta[which(meta$id %in% rownames(y5)),2:5])
    y5$coord <- chr7.cov5$coord[x]
   y6 <- chr7.cov5.meta[x,9:19]
   y7<- colnames(y6)[which(!is.na(y6))]
    df<- y5[which(y5$embryo %in% y7),]
     
     return(df)
     
  }))
 
gg1 <- ggplot(snp.chr7.cov5[which(snp.chr7.cov5$cluster=='0'),], aes(x=ref, y=alt,color= cluster)) +
  geom_point()+
  geom_abline(intercept = 0, slope = 1,color="grey85")+
  theme_bw()+labs(title="mean of SNPs per cell-mean.chr7.cov5")+ xlim(0, 8)+ ylim(0, 8)
gg2 <- ggplot(snp.chr7.cov5[which(snp.chr7.cov5$cluster=='1'),], aes(x=ref, y=alt,color= cluster)) +
  geom_point()+
  geom_abline(intercept = 0, slope = 1,color="grey85")+
  theme_bw()+labs(title="mean of SNPs per cell-mean.chr7.cov5")+ xlim(0, 8)+ ylim(0, 8)
gg3 <- ggplot(snp.chr7.cov5[which(snp.chr7.cov5$cluster=='2'),], aes(x=ref, y=alt,color= cluster)) +
  geom_point()+
  geom_abline(intercept = 0, slope = 1,color="grey85")+
  theme_bw()+labs(title="mean of SNPs per cell-mean.chr7.cov5")+ xlim(0, 8)+ ylim(0, 8)
gg4 <- ggplot(snp.chr7.cov5[which(snp.chr7.cov5$cluster=='3'),], aes(x=ref, y=alt,color= cluster)) +
  geom_point()+
  geom_abline(intercept = 0, slope = 1,color="grey85")+
  theme_bw()+labs(title="mean of SNPs per cell-mean.chr7.cov5")+ xlim(0, 8)+ ylim(0, 8)
gg5 <- ggplot(snp.chr7.cov5[which(snp.chr7.cov5$cluster=='4'),], aes(x=ref, y=alt,color= cluster)) +
  geom_point()+
  geom_abline(intercept = 0, slope = 1,color="grey85")+
  theme_bw()+labs(title="mean of SNPs per cell-mean.chr7.cov5")+ xlim(0, 8)+ ylim(0, 8)

ggarrange(gg1,gg2,gg3,gg4,gg5, ncol = 2,nrow = 3)

###################################chrX-pro#####################################

chrx.ref <- xnorm[,grep(".1$", colnames(xnorm))]
chrx.alt <- xnorm[,grep(".2$", colnames(xnorm))]

chrx.pct <- chrx.ref+chrx.alt

for (i in 1:nrow(chrx.pct)) {
  y1 <- chrx.pct[i,]
  y2 <- which(y1 <15)
  chrx.ref[i,y2] <- NA
  chrx.alt[i,y2] <- NA
}

chrx.ref <- log(chrx.ref+1)
chrx.alt <- log(chrx.alt+1)

chrx.ref <- data.frame(chrx.cov5[,1:9],chrx.ref)
chrx.alt <- data.frame(chrx.cov5[,1:9],chrx.alt)

ref.xpro <- chrx.ref[-which(chrx.ref$hgnc %in% Xesclist),]
alt.xpro <- chrx.alt[-which(chrx.alt$hgnc %in% Xesclist),]

ref<- ref.xpro[,10:236]
alt <- alt.xpro[,10:236]

##### 
y4 <- as.data.frame(str_match(colnames(ref), "(.*)_(.*)_CL_(.*)$"))
y4$cluster <- stringr::str_replace(y4$V4, '.\\d+', '')
y4$V4 <- c("ref")
colnames(y4)<- c("id","embryo","cell","allele","cluster")

y5 <- as.data.frame(str_match(colnames(alt), "(.*)_(.*)_CL_(.*)$"))
y5$cluster <- stringr::str_replace(y5$V4, '.\\d+', '')
y5$V4 <- c("alt")
colnames(y5)<- c("id","embryo","cell","allele","cluster")

meta <- rbind(y4,y5)

 snp.xpro.cov5 <- do.call(rbind,lapply(1:nrow(ref), function(x){
   
   y3 <- ref[x,]
   y4<- alt[x,]
   y5 <- as.data.frame(cbind(colSums(y3),colSums(y4)))
   colnames(y5) <- c("ref","alt")
   y5 <- data.frame(y5,meta[which(meta$id %in% rownames(y5)),2:5])
    y5$coord <- meta.xpro$coord[x]
   y6 <- meta.xpro[x,10:20]
   y7<- colnames(y6)[which(!is.na(y6))]
    df<- y5[which(y5$embryo %in% y7),]
     
     return(df)
     
  }))

gg1 <- ggplot(snp.xpro.cov5[which(snp.xpro.cov5$cluster=='0'),], aes(x=ref, y=alt,color= cluster)) +
  geom_point()+
  geom_abline(intercept = 0, slope = 1,color="grey85")+
  theme_bw()+labs(title="snp.xpro.cov5")+ xlim(0, 8)+ ylim(0, 8)
gg2 <- ggplot(snp.xpro.cov5[which(snp.xpro.cov5$cluster=='1'),], aes(x=ref, y=alt,color= cluster)) +
  geom_point()+
  geom_abline(intercept = 0, slope = 1,color="grey85")+
  theme_bw()+labs(title="snp.xpro.cov5")+ xlim(0, 8)+ ylim(0, 8)
gg3 <- ggplot(snp.xpro.cov5[which(snp.xpro.cov5$cluster=='2'),], aes(x=ref, y=alt,color= cluster)) +
  geom_point()+
  geom_abline(intercept = 0, slope = 1,color="grey85")+
  theme_bw()+labs(title="snp.xpro.cov5")+ xlim(0, 8)+ ylim(0, 8)
gg4 <- ggplot(snp.xpro.cov5[which(snp.xpro.cov5$cluster=='3'),], aes(x=ref, y=alt,color= cluster)) +
  geom_point()+
  geom_abline(intercept = 0, slope = 1,color="grey85")+
  theme_bw()+labs(title="snp.xpro.cov5")+ xlim(0, 8)+ ylim(0, 8)
gg5 <- ggplot(snp.xpro.cov5[which(snp.xpro.cov5$cluster=='4'),], aes(x=ref, y=alt,color= cluster)) +
  geom_point()+
  geom_abline(intercept = 0, slope = 1,color="grey85")+
  theme_bw()+labs(title="snp.xpro.cov5")+ xlim(0, 8)+ ylim(0, 8)

ggarrange(gg1,gg2,gg3,gg4,gg5, ncol = 2,nrow = 3)
```


##Fig.S4C chrX-esc scatter plot
```{r, echo=TRUE, error=TRUE}
ref.xesc <- chrx.ref[which(chrx.ref$hgnc %in% Xesclist),]
alt.xesc <- chrx.alt[which(chrx.alt$hgnc %in% Xesclist),]

ref<- ref.xesc[,10:236]
alt <- alt.xesc[,10:236]

##### 
y4 <- as.data.frame(str_match(colnames(ref), "(.*)_(.*)_CL_(.*)$"))
y4$cluster <- stringr::str_replace(y4$V4, '.\\d+', '')
y4$V4 <- c("ref")
colnames(y4)<- c("id","embryo","cell","allele","cluster")

y5 <- as.data.frame(str_match(colnames(alt), "(.*)_(.*)_CL_(.*)$"))
y5$cluster <- stringr::str_replace(y5$V4, '.\\d+', '')
y5$V4 <- c("alt")
colnames(y5)<- c("id","embryo","cell","allele","cluster")

meta <- rbind(y4,y5)

 snp.xesc.cov5 <- do.call(rbind,lapply(1:nrow(ref), function(x){
   
   y3 <- ref[x,]
   y4<- alt[x,]
   y5 <- as.data.frame(cbind(colSums(y3),colSums(y4)))
   colnames(y5) <- c("ref","alt")
   y5 <- data.frame(y5,meta[which(meta$id %in% rownames(y5)),2:5])
    y5$coord <- meta.xesc$coord[x]
   y6 <- meta.xesc[x,10:20]
   y7<- colnames(y6)[which(!is.na(y6))]
    df<- y5[which(y5$embryo %in% y7),]
     
     return(df)
     
  }))


gg1 <- ggplot(snp.xesc.cov5[which(snp.xesc.cov5$cluster=='0'),], aes(x=ref, y=alt,color= cluster)) +
  geom_point()+
  geom_abline(intercept = 0, slope = 1,color="grey85")+
  theme_bw()+labs(title="snp.xesc.cov5")+ xlim(0, 8)+ ylim(0, 8)
gg2 <- ggplot(snp.xesc.cov5[which(snp.xesc.cov5$cluster=='1'),], aes(x=ref, y=alt,color= cluster)) +
  geom_point()+
  geom_abline(intercept = 0, slope = 1,color="grey85")+
  theme_bw()+labs(title="snp.xesc.cov5")+ xlim(0, 8)+ ylim(0, 8)
gg3 <- ggplot(snp.xesc.cov5[which(snp.xesc.cov5$cluster=='2'),], aes(x=ref, y=alt,color= cluster)) +
  geom_point()+
  geom_abline(intercept = 0, slope = 1,color="grey85")+
  theme_bw()+labs(title="snp.xesc.cov5")+ xlim(0, 8)+ ylim(0, 8)
gg4 <- ggplot(snp.xesc.cov5[which(snp.xesc.cov5$cluster=='3'),], aes(x=ref, y=alt,color= cluster)) +
  geom_point()+
  geom_abline(intercept = 0, slope = 1,color="grey85")+
  theme_bw()+labs(title="snp.xesc.cov5")+ xlim(0, 8)+ ylim(0, 8)
gg5 <- ggplot(snp.xesc.cov5[which(snp.xesc.cov5$cluster=='4'),], aes(x=ref, y=alt,color= cluster)) +
  geom_point()+
  geom_abline(intercept = 0, slope = 1,color="grey85")+
  theme_bw()+labs(title="snp.xesc.cov5")+ xlim(0, 8)+ ylim(0, 8)

ggarrange(gg1,gg2,gg3,gg4,gg5, ncol = 2,nrow = 3)
    
```


##Fig.S4E
```{r, echo=TRUE, error=TRUE}
#############chromosome 7#######################
chr7.ref <- chr7.cov5.2[,grep(".1$", colnames(chr7.cov5.2))]
chr7.alt <- chr7.cov5.2[,grep(".2$", colnames(chr7.cov5.2))]

chr7.pct <- chr7.ref+chr7.alt

for (i in 1:nrow(chr7.pct)) {
  y1 <- chr7.pct[i,]
  y2 <- which(y1 <15)
  chr7.pct[i,y2] <- NA
}

chr7.ref.pct<- chr7.ref/chr7.pct
chr7.alt.pct<- 1-chr7.ref.pct

chr7.ref.pct <- data.frame(chr7.cov5.2[,c(1:9)], chr7.ref.pct)
chr7.alt.pct <- data.frame(chr7.cov5.2[,c(1:9)], chr7.alt.pct)

ref<- chr7.ref.pct[,10:236]
alt <- chr7.alt.pct[,10:236]

chr7.pct2 <- abs(ref-alt)  ##percentage ranking from 0 to 1

##### 
meta <- as.data.frame(str_match(colnames(chr7.pct2), "(.*)_(.*)_CL_(.*)$"))
meta$cluster <- stringr::str_replace(meta$V4, '.\\d+', '')
meta$V4 <- c("ref")
colnames(meta)<- c("id","embryo","cell","allele","cluster")




 snp.chr7 <- do.call(rbind,lapply(1:nrow(ref), function(x){
   
   y3 <- chr7.pct2[x,]
   y4 <- data.frame(colSums(y3, na.rm=TRUE),meta[,2:5])
    y4$coord <- chr7.cov5.meta$coord[x]
   y5 <- chr7.cov5.meta[x,9:19]
   y6<- colnames(y5)[which(!is.na(y5))]
    df<- y4[which(y4$embryo %in% y6),]
     
     return(df)
     
  }))

colnames(snp.chr7)[1] <- "percentage"

#############chromosome X########################
chrx.ref <- xnorm[,grep(".1$", colnames(xnorm))]
chrx.alt <- xnorm[,grep(".2$", colnames(xnorm))]

chrx.pct <- chrx.ref+chrx.alt

for (i in 1:nrow(chrx.pct)) {
  y1 <- chrx.pct[i,]
  y2 <- which(y1 <15)
  chrx.pct[i,y2] <- NA
}

chrx.ref.pct<- chrx.ref/chrx.pct
chrx.alt.pct<- 1-chrx.ref.pct

chrx.ref.pct <- data.frame(xnorm[,c(1:9)], chrx.ref.pct)
chrx.alt.pct <- data.frame(xnorm[,c(1:9)], chrx.alt.pct)

xpro.ref.pct<- chrx.ref.pct [-which(chrx.ref.pct$hgnc %in% Xesclist),]
xpro.alt.pct<- chrx.alt.pct [-which(chrx.alt.pct$hgnc %in% Xesclist),]

xesc.ref.pct<- chrx.ref.pct [which(chrx.ref.pct$hgnc %in% Xesclist),]
xesc.alt.pct<- chrx.alt.pct [which(chrx.alt.pct$hgnc %in% Xesclist),]

##################xpro#########################################
ref<- xpro.ref.pct[,10:236]
alt <- xpro.alt.pct[,10:236]

xpro.pct2 <- abs(ref-alt)  ##percentage ranking from 0 to 1

##### 
meta <- as.data.frame(str_match(colnames(xpro.pct2), "(.*)_(.*)_CL_(.*)$"))
meta$cluster <- stringr::str_replace(meta$V4, '.\\d+', '')
meta$V4 <- c("ref")
colnames(meta)<- c("id","embryo","cell","allele","cluster")


 snp.xpro <- do.call(rbind,lapply(1:nrow(ref), function(x){
   
   y3 <- xpro.pct2[x,]
   y4 <- data.frame(colSums(y3, na.rm=TRUE),meta[,2:5])
    y4$coord <- meta.xpro$coord[x]
   y5 <- meta.xpro[x,10:20]
   y6<- colnames(y5)[which(!is.na(y5))]
    df<- y4[which(y4$embryo %in% y6),]
     
     return(df)
     
  }))

colnames(snp.xpro)[1] <- "percentage"

#################xesc##########################################
ref<- xesc.ref.pct[,10:236]
alt <- xesc.alt.pct[,10:236]

xesc.pct2 <- abs(ref-alt)  ##percentage ranking from 0 to 1

##### 
meta <- as.data.frame(str_match(colnames(xesc.pct2), "(.*)_(.*)_CL_(.*)$"))
meta$cluster <- stringr::str_replace(meta$V4, '.\\d+', '')
meta$V4 <- c("ref")
colnames(meta)<- c("id","embryo","cell","allele","cluster")


 snp.xesc <- do.call(rbind,lapply(1:nrow(ref), function(x){
   
   y3 <- xesc.pct2[x,]
   y4 <- data.frame(colSums(y3, na.rm=TRUE),meta[,2:5])
    y4$coord <- meta.xesc$coord[x]
   y5 <- meta.xesc[x,10:20]
   y6<- colnames(y5)[which(!is.na(y5))]
    df<- y4[which(y4$embryo %in% y6),]
     
     return(df)
     
  }))

colnames(snp.xesc)[1] <- "percentage"

####################ggplot###################################
snp.chr7$group <- "chr7"
snp.xpro$group <- "xpro"
snp.xesc$group <- "xesc"

snp.pct <- rbind(snp.chr7,snp.xpro,snp.xesc)

snp.pct$cluster <- factor(snp.pct$cluster,levels = c("0","1","4","3","2"))

ggplot(snp.pct, aes(x=cluster, y=percentage,fill=group)) + 
  geom_boxplot(outlier.size=0.2)+ theme_bw()+
  scale_fill_brewer(palette="Set2")

```


##Fig.S4F
```{r, echo=TRUE, error=TRUE}
snp.xpro$type <- "monoallelic"
snp.xpro[which(snp.xpro$percentage<0.6),8]<-"biallelic"

for(i in unique(snp.xpro$coord)){
  y1 <- snp.xpro[which(snp.xpro$coord==i),]
  if(length(unique(y1$cluster))!=5){
    snp.xpro[which(snp.xpro$coord==i),]<-NA
  }
  }

snp.xpro <- snp.xpro[complete.cases(snp.xpro), ]
y1 <- snp.xpro[which(snp.xpro$cluster=="0"),]
length(unique(y1$coord))

for(i in unique(y1$coord)){
  y2 <- y1[which(y1$coord==i),]
  y3 <- y2[which(y2$type=="biallelic"),]
  if(nrow(y3)/nrow(y2)<0.8){
    y1[which(y1$coord==i),]<-NA
  }
  
  }

y2 <- y1[complete.cases(y1),]

y3<-unique(y2$coord)
y4 <-snp.xpro[which(snp.xpro$coord %in% y3),]
y4$plot <- "xpro"
length(unique(y4$coord))


y4$cellid <- y4$embryo
y4$cellid <- paste(y4$cellid,y4$cell)

#####calculate biallelic/monoallelic percentage
pct.xpro <- do.call(rbind,lapply(unique(y4$cellid), function(i){
  y2 <- y4[which(y4$cellid==i),]
  y3 <- length(which(y2$type=="biallelic"))/nrow(y2)
  y5 <- length(which(y2$type=="monoallelic"))/nrow(y2)
  y6 <- list(y3,y5)
  y6 <- as.data.frame(y6)
  colnames(y6)<- c("biallelic","monoallelic")
  y6$cellid <- i
  y6$cluster <- unique(y2$cluster)
  
  return(y6)
  
  }))

pct.xpro.mono <- pct.xpro[,2:4]

pct.xpro.mono$cluster <- factor(pct.xpro.mono$cluster,levels = c("0","1","4","3","2"))
ggplot(pct.xpro.mono, aes(x=cluster, y=monoallelic)) + 
  geom_boxplot(outlier.size=0.2)+ theme_bw()+
  scale_fill_brewer(palette="Set2")

```



##recombine female male
```{r, echo=TRUE, error=TRUE}
maleMeiosis <- subset(male, ident=c(1,2,3,5,6,7))
maleMeiosis <- RenameIdents(object = maleMeiosis, 
                            '1' = 'M_Z',
                            '2'= 'M_preL',
                            '3' = 'M_L', 
                            '5' = 'M_lP',
                            '6' = 'M_D',
                            '7' = 'M_eP')

femaleMeiosis <- lateMeiot
femaleMeiosis <- RenameIdents(object = femaleMeiosis,
                          '0' = 'F_preL',
                          '1' = 'F_L',
                          '4' = 'F_Z',
                          '3' = 'F_P',
                          '2' = 'F_D')

## Combine datasets
maleFemaleMeiosis <- merge(maleMeiosis, y=femaleMeiosis, add.cell.ids = c("M_", "F_"))

## Cells in each dataset
dim(maleMeiosis)
dim(femaleMeiosis)

maleFemaleMeiosis <- FindVariableFeatures(object = maleFemaleMeiosis, 
                         mean.function = ExpMean, 
                         dispersion.function = LogVMR,
                         nfeatures = 2500)
length(x = VariableFeatures(maleFemaleMeiosis))

#####scaling the data
maleFemaleMeiosis <- ScaleData(object = maleFemaleMeiosis)

#####PCA
maleFemaleMeiosis <- RunPCA(object = maleFemaleMeiosis, 
              features = VariableFeatures(maleFemaleMeiosis), 
              do.print = TRUE, 
              pcs.print = 1:5, 
              genes.print = 5)

VizDimLoadings(object = maleFemaleMeiosis, 
               dims = 1:2, 
               reduction = "pca")

DimPlot(object = maleFemaleMeiosis, reduction = "pca")

ElbowPlot(object = maleFemaleMeiosis)

## Print the percenatge of the standard deviation of each PC as a 
## fraction of the total standard deviation od the first 20 PCs
totalSdev <- sum(Stdev(object = maleFemaleMeiosis, reduction = "pca"))
print(Stdev(object = maleFemaleMeiosis, reduction = "pca")/totalSdev)

maleFemaleMeiosis <- FindNeighbors(maleFemaleMeiosis,
                                   reduction = "pca",
                                   dims = 1:11)
maleFemaleMeiosis <- FindClusters(object = maleFemaleMeiosis, 
                                  resolution = 0.6)

maleFemaleMeiosis <- RunTSNE(object = maleFemaleMeiosis, 
                             dims = 1:11)
maleFemaleMeiosis <- RunUMAP(object = maleFemaleMeiosis, 
                             dims = 1:11)

DimPlot(object = maleFemaleMeiosis, 
        reduction = "tsne")

###Get gender of cells and plot
gender <- sapply(colnames(x = maleFemaleMeiosis), function(x){
  strsplit(x, split = "__", fixed = TRUE)[[1]][1]
})
maleFemaleMeiosis$gender <- gender
DimPlot(object = maleFemaleMeiosis, 
             reduction = "tsne",
             group.by = "gender")



####randomly select 230 male cells
set.seed(42)
sampledCellIDs <- sample(x = colnames(maleMeiosis), size = 230, replace = FALSE)
maleMeiosisSubsmpl <- subset(x = maleMeiosis, cells=sampledCellIDs)
maleFemaleMeiosisEqualized <- merge(maleMeiosisSubsmpl, y=femaleMeiosis, add.cell.ids = c("M_", "F_"))
## Add the M_ in the cell ID
sampledIDs <- paste0("M__",sampledCellIDs)
DimPlot(object = maleFemaleMeiosis, 
             reduction = "tsne",
             cells.highlight = sampledIDs)


###Find Variable Genes
maleFemaleMeiosisEqualized <- FindVariableFeatures(object = maleFemaleMeiosisEqualized, 
                         mean.function = ExpMean, 
                         dispersion.function = LogVMR,
                         nfeatures = 2500)
length(x = VariableFeatures(maleFemaleMeiosisEqualized))

topAveExpr = HVFInfo(maleFemaleMeiosisEqualized)[HVFInfo(maleFemaleMeiosisEqualized)[,1]>3, ]

##Scaling the data
maleFemaleMeiosisEqualized <- ScaleData(object = maleFemaleMeiosisEqualized)
##PCA
maleFemaleMeiosisEqualized <- RunPCA(object = maleFemaleMeiosisEqualized, 
                                     features = VariableFeatures(maleFemaleMeiosisEqualized), 
                                     do.print = TRUE, 
                                     pcs.print = 1:5, 
                                     genes.print = 5)

##Vizualize PCA loadings
VizDimLoadings(object = maleFemaleMeiosisEqualized, 
               dims = 1:2, 
               reduction = "pca")

DimPlot(object = maleFemaleMeiosisEqualized, reduction = "pca")

ElbowPlot(object = maleFemaleMeiosisEqualized)


###tSNEs and UMAP
maleFemaleMeiosisEqualized <- RunTSNE(object = maleFemaleMeiosisEqualized, 
                             dims = 1:11)
maleFemaleMeiosisEqualized <- RunUMAP(object = maleFemaleMeiosisEqualized, 
                             dims = 1:11)
DimPlot(object = maleFemaleMeiosisEqualized, 
        reduction = "tsne")

```



##Fig.4G
```{r, echo=TRUE, error=TRUE}
##import table
ubi<- read_excel("C:/Users/fxy19/Desktop/mmc4.xlsx")

################################################calculate femalemale dataset################################
genes <- GetAssayData(object = maleFemaleMeiosisEqualized, slot = "data")
dim(genes)

ubi.FM <- genes[which(rownames(genes) %in% ubi$gene_short_name),]

####################


ubi.FMa <- getBM(attributes = c('hgnc_symbol', 'chromosome_name'),
                   filters = 'hgnc_symbol', 
                   values = rownames(ubi.FM), 
                   mart = mart)

ubi.FMa2<- ubi.FM[-which(rownames(ubi.FM) %in% ubi.FMa$hgnc_symbol),]
dim(ubi.FMa2)
ubi.FMa2 <- getBM(attributes = c('hgnc_symbol', 'chromosome_name','external_synonym'),
                    filters = 'external_synonym', 
                    values = rownames(ubi.FMa2), 
                    mart = mart)
ubi.FMa2 $hgnc_symbol <- ubi.FMa2$external_synonym
ubi.FMa2<- ubi.FMa2[,-3]

ubi.genes <- rbind(ubi.FMa,ubi.FMa2)
ubi.genes <- ubi.genes[-c(grep("CHR_",ubi.genes$chromosome_name)),]
####8050 genes
chr.A <- ubi.genes[which(ubi.genes$chromosome_name %in% as.character(c(1:22))) ,]
chr.X<- ubi.genes[which(ubi.genes$chromosome_name =="X"),]
chr.7<- ubi.genes[which(ubi.genes$chromosome_name =="7"),]
######
genes.A <- genes[which(rownames(genes) %in% chr.A$hgnc_symbol),]
genes.X<- genes[which(rownames(genes) %in% chr.X$hgnc_symbol),]
genes.7<- genes[which(rownames(genes) %in% chr.7$hgnc_symbol),]

genes.Xesc <- genes.X[which(rownames(genes.X) %in% Xesclist),]
genes.Xpro <- genes.X[!(rownames(genes.X) %in% Xesclist), ]

nrow(genes.A)
nrow(genes.7)
nrow(genes.X)
nrow(genes.Xesc)
nrow(genes.Xpro)

##############################

## calculate A-linked genes  mean expression per cell
genes.A.t<- data.frame(t(as.matrix(genes.A)))

meanA<- apply(genes.A.t, 1, FUN = mean)
genes.A.t$mean <- meanA
genes.A.t<-as.matrix(genes.A.t)
meanA<-as.matrix(genes.A.t[,-(1:7793)])

## calculate X-esc genes  mean expression per cell
genes.Xesc.t<- data.frame(t(as.matrix(genes.Xesc)))

meanXesc<- apply(genes.Xesc.t, 1, FUN = mean)
genes.Xesc.t$mean <- meanXesc
genes.Xesc.t<-as.matrix(genes.Xesc.t)
meanXesc<-as.matrix(genes.Xesc.t[,-(1:32)])

## calculate X-sub genes  mean expression per cell
genes.Xpro.t<- data.frame(t(as.matrix(genes.Xpro)))

meanXpro<- apply(genes.Xpro.t, 1, FUN = mean)
genes.Xpro.t$mean <- meanXpro
genes.Xpro.t<-as.matrix(genes.Xpro.t)
meanXpro<-as.matrix(genes.Xpro.t[,-(1:219)])

## calculate chr7 genes  mean expression per cell
genes.chr7.t<- data.frame(t(as.matrix(genes.7)))

meanchr7<- apply(genes.chr7.t, 1, FUN = mean)
genes.chr7.t$mean <- meanchr7
genes.chr7.t<-as.matrix(genes.chr7.t)
meanchr7<-as.matrix(genes.chr7.t[,-(1:378)])

ratiomean<- data.frame( meanA, meanXesc, meanXpro, meanchr7)

## calculate X:A ratio per cell
Xescratio <- ratiomean$meanXesc/ratiomean$meanA
ratiomean$Xescratio <- Xescratio
Xproratio <- ratiomean$meanXpro/ratiomean$meanA
ratiomean$Xproratio <- Xproratio
chr7ratio <- ratiomean$meanchr7/ratiomean$meanA
ratiomean$chr7ratio <- chr7ratio

###############draw boxplot #####################

groupIdentity <- Idents(object = maleFemaleMeiosisEqualized)
ratiomean$groupIdentity <- groupIdentity
ratiomean<- as.data.frame(ratiomean)

df <- ratiomean[,5:8]
df <- melt(df, ID="groupIdentity")
df$groupIdentity <- factor(df$groupIdentity,levels = c("F_preL","F_L","F_Z","F_P","F_D","M_preL","M_L","M_Z","M_eP","M_lP","M_D"))

ggplot(df, aes(x = groupIdentity, y = value, fill=variable)) + 
  geom_boxplot(outlier.size=0.2)+ theme_bw()+
  scale_fill_brewer(palette="Set2")+ylim(0,2)

```



##Fig.S4D 
```{r, echo=TRUE, error=TRUE}
genes <- GetAssayData(object = maleFemaleMeiosisEqualized, slot = "data")
dim(genes)
chr.genes <- getBM(attributes = c('hgnc_symbol', 'chromosome_name'),
                                filters = 'hgnc_symbol', 
                                values = rownames(genes), 
                                mart = mart)

chr.genes2 <- genes[-which(rownames(genes) %in% chr.genes$hgnc_symbol),]
dim(chr.genes2)
chr.genes2 <- getBM(attributes = c('hgnc_symbol', 'chromosome_name','external_synonym'),
                                filters = 'external_synonym', 
                                values = rownames(chr.genes2), 
                                mart = mart)
chr.genes2 $hgnc_symbol <- chr.genes2$external_synonym
chr.genes2 <- chr.genes2[,-3]

chr.genes <- rbind(chr.genes,chr.genes2)
chr.genes <- chr.genes[-c(grep("CHR_",chr.genes$chromosome_name)),]

chr.X<- chr.genes[which(chr.genes$chromosome_name =="X"),]
chr.A <- chr.genes[which(chr.genes$chromosome_name %in% as.character(c(1:22))) ,]
chr.7<- chr.genes[which(chr.genes$chromosome_name =="7"),]
######

genes.A <- genes[which(rownames(genes) %in% chr.A$hgnc_symbol),]
genes.X<- genes[which(rownames(genes) %in% chr.X$hgnc_symbol),]
genes.7<- genes[which(rownames(genes) %in% chr.7$hgnc_symbol),]

genes.Xesc <- genes.X[which(rownames(genes.X) %in% Xesclist),]
genes.Xpro <- genes.X[!(rownames(genes.X) %in% Xesclist), ]

nrow(genes.A)
nrow(genes.7)
nrow(genes.X)
nrow(genes.Xesc)
nrow(genes.Xpro)

##############################

## calculate A-linked genes  mean expression per cell
genes.A.t<- data.frame(t(as.matrix(genes.A)))

meanA<- apply(genes.A.t, 1, FUN = mean)
genes.A.t$mean <- meanA
genes.A.t<-as.matrix(genes.A.t)
meanA<-as.matrix(genes.A.t[,-(1:20179)])

## calculate X-esc genes  mean expression per cell
genes.Xesc.t<- data.frame(t(as.matrix(genes.Xesc)))

meanXesc<- apply(genes.Xesc.t, 1, FUN = mean)
genes.Xesc.t$mean <- meanXesc
genes.Xesc.t<-as.matrix(genes.Xesc.t)
meanXesc<-as.matrix(genes.Xesc.t[,-(1:92)])

## calculate X-sub genes  mean expression per cell
genes.Xpro.t<- data.frame(t(as.matrix(genes.Xpro)))

meanXpro<- apply(genes.Xpro.t, 1, FUN = mean)
genes.Xpro.t$mean <- meanXpro
genes.Xpro.t<-as.matrix(genes.Xpro.t)
meanXpro<-as.matrix(genes.Xpro.t[,-(1:697)])

## calculate chr7 genes  mean expression per cell
genes.chr7.t<- data.frame(t(as.matrix(genes.7)))

meanchr7<- apply(genes.chr7.t, 1, FUN = mean)
genes.chr7.t$mean <- meanchr7
genes.chr7.t<-as.matrix(genes.chr7.t)
meanchr7<-as.matrix(genes.chr7.t[,-(1:994)])

ratiomean<- data.frame( meanA, meanXesc, meanXpro, meanchr7)

## calculate X:A ratio per cell
Xescratio <- ratiomean$meanXesc/ratiomean$meanA
ratiomean$Xescratio <- Xescratio
Xproratio <- ratiomean$meanXpro/ratiomean$meanA
ratiomean$Xproratio <- Xproratio
chr7ratio <- ratiomean$meanchr7/ratiomean$meanA
ratiomean$chr7ratio <- chr7ratio

###############draw boxplot #####################

groupIdentity <- Idents(object = maleFemaleMeiosisEqualized)
ratiomean$groupIdentity <- groupIdentity
ratiomean<- as.data.frame(ratiomean)

df <- ratiomean[,5:8]
df <- melt(df, ID="groupIdentity")
df$groupIdentity <- factor(df$groupIdentity,levels = c("F_preL","F_L","F_Z","F_P","F_D","M_preL","M_L","M_Z","M_eP","M_lP","M_D"))

ggplot(df, aes(x = groupIdentity, y = value, fill=variable)) + 
  geom_boxplot(outlier.size=0.2)+ theme_bw()+
  scale_fill_brewer(palette="Set2")+ylim(0,2)

```


##Fig.4H 
```{r, echo=TRUE, error=TRUE}

L <- c("RNF8","TOPBP1","MDC1","FAAP24","ATR","H2AFX","BRCA1","HDAC11","HDAC9","HDAC8","HDAC2","HDAC1","KDM5C","KDM5B","KDM5A","SETD8","SETDB1","EHMT2","EHMT1","JARID2","AEBP2","EED","SUZ12","EZH2","EZH1","KDM7A","PHF8","KDM6B","KDM6A","KDM4D","KDM4C","KDM4B","KDM4A","KDM3B","KDM3A","EP300","CREBBP","SETD2","KMT2C","KMT2B","KMT2A","SETD1A")

DotPlot(maleFemaleMeiosisEqualized, features = L , cols = c("blue", "red"), dot.scale = 6)+ RotatedAxis()

```

